Allow specifying ExternalTrafficPolicy for Services with ExternalIPs

tnqn opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When defining a Service, we can specify externalIP for any Service type, and the traffic policy of externalIP is subject to externalTrafficPolicy. However, the policy can't be set when type is not NodePort or LoadBalancer, and will default to Cluster when kube-proxy processes the Service.

For example, creating the following Service will fail:
```
kubectl apply -f - <<EOF
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app.kubernetes.io/name: MyApp
  ports:
    - name: http
      protocol: TCP
      port: 80
      targetPort: 49152
  externalIPs:
    - 198.51.100.32
  externalTrafficPolicy: Local
EOF

---
The Service "my-service" is invalid: spec.externalTrafficPolicy: Invalid value: "Local": may only be set when `type` is 'NodePort' or 'LoadBalancer'
```

This commit updates the defaulting and validation of Service to allow specifying ExternalTrafficPolicy when ExternalIP is not empty.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Allow specifying ExternalTrafficPolicy for Services with ExternalIPs.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

tnqn committed:
Allow specifying ExternalTrafficPolicy for ClusterIP Services with ExternalIPs

When defining a ClusterIP Service, we can specify externalIP, and the
traffic policy of externalIP is subject to externalTrafficPolicy.
However, the policy can't be set when type is not NodePort or
LoadBalancer, and will default to Cluster when kube-proxy processes the
Service.

This commit updates the defaulting and validation of Service to allow
specifying ExternalTrafficPolicy for ClusterIP Services with
ExternalIPs.

Signed-off-by: Quan Tian <qtian@vmware.com>

tnqn commented:
/sig network

danwinship commented:
I'm fine answering either "yes, that definitely makes sense as an edge case we should have supported all along" or "we're trying to avoid further combinations of features in Service so have you considered trying to implement this via Gateway?" :slightly_smiling_face:

aojea commented:
I forget to mention that ExternalIP has also this CVE https://github.com/kubernetes/kubernetes/issues/97076

thockin commented:
ACK the corner-case.  I also agree with Dan that it's not obvious we want to fix this.  It's more brittle than anyone is happy about.  The PR seems simple, but I have to think about it a bit to see if we're missing something.  We always are.

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -125,7 +125,8 @@ func SetDefaults_Service(obj *v1.Service) {
 	// to Global for consistency.
 	if (obj.Spec.Type == v1.ServiceTypeNodePort ||
-		obj.Spec.Type == v1.ServiceTypeLoadBalancer) &&
+		obj.Spec.Type == v1.ServiceTypeLoadBalancer ||
```
I think this wants to be `NodePort || LB || (ClusterIP && externalIPs)`, doesn't it?  We don't want to support eTP on `type=ExternalName`

thockin replied:
externalIPs being set when type=ExternalName is a bug.  Proxies skip it.

thockin commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -1563,6 +1563,13 @@ func TestSetDefaultServiceExternalTraffic(t *testing.T) {
+	out = obj.(*v1.Service)
+	if out.Spec.ExternalTrafficPolicy != v1.ServiceExternalTrafficPolicyCluster {
+		t.Errorf("Expected ExternalTrafficPolicy to be %v, got %v", v1.ServiceExternalTrafficPolicyCluster, out.Spec.ExternalTrafficPolicy)
+	}
```
You should add a case for `Type: "ExternalName"` which fails.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5122,7 +5122,7 @@ func validateServicePort(sp *core.ServicePort, requireName, isHeadlessService bo
 
 func needsExternalTrafficPolicy(svc *core.Service) bool {
-	return svc.Spec.Type == core.ServiceTypeLoadBalancer || svc.Spec.Type == core.ServiceTypeNodePort
+	return svc.Spec.Type == core.ServiceTypeLoadBalancer || svc.Spec.Type == core.ServiceTypeNodePort || len(svc.Spec.ExternalIPs) > 0
```
Same as above: `NodePort || LB || (ClusterIP && externalIPs)`

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -14050,25 +14050,38 @@ func TestValidateServiceCreate(t *testing.T) {
 			s.Spec.ExternalIPs = []string{"127.0.0.1"}
 		},
-		numErrs: 1,
+		numErrs: 2,
```
I assume this is the additional "required" error?  We should adjust the test loop to set this properly, so each case is really testing just one thing

tnqn replied:
The new "required" was externalTrafficPolicy is required when ExternalIP is set for ClusterIP service.
I have adjusted the new tests to test one thing only.

thockin commented on pkg/api/v1/service/util.go:
```diff
@@ -70,7 +70,8 @@ func GetLoadBalancerSourceRanges(service *v1.Service) (utilnet.IPNetSet, error)
 func ExternalPolicyLocal(service *v1.Service) bool {
 	if service.Spec.Type != v1.ServiceTypeLoadBalancer &&
-		service.Spec.Type != v1.ServiceTypeNodePort {
+		service.Spec.Type != v1.ServiceTypeNodePort &&
```
Same: `NodePort || LB || (ClusterIP && externalIPs)`

thockin reviewed (commented):
I'm still not sure about this, but first-pass feedback

tnqn commented:
@aojea @danwinship @thockin thanks for your comments. I thought this is a neglected case so created the PR to fix it. I'm fine with closing it if we just want to keep it as is, or further enhancing the validation in the way you think making most sense. Anyway I have addressed the current comments.

> how is this going to affect kube-proxy?

Currently if we create a ClusterIP service with ExternalIPs, kube-proxy will just use Cluster externalTrafficPolicy to realize ExternalIPs as the policy is not allowed to be set.  With this commit, the policy can be set to Cluster or Local, and kube-proxy can realize either of the policies for ExternalIPs of ClusterIP service.

> I'm fine answering either "yes, that definitely makes sense as an edge case we should have supported all along" or "we're trying to avoid further combinations of features in Service so have you considered trying to implement this via Gateway?"

Do you mean the combination of "ClusterIP Service + ExternalIPs" or "ClusterIP Service + ExternalIPs + ExternalTrafficLocal"? The former seems an example of ExternalIPs in https://kubernetes.io/docs/concepts/services-networking/service/#external-ips.

> I forget to mention that ExternalIP has also this CVE #97076

Thanks for the pointer. It seems the CVE is about ExternalIP itself, not whether its traffic policy can be set to Local, do you think this change could make it worse?

aojea commented:
>  It seems the CVE is about ExternalIP itself, not whether its traffic policy can be set to Local, do you think this change could make it worse?

I'm a bit busy with code freeze coming, sorry for not being able to spend more time,  I just wanted to say that we are "encouraging" to not use ExternalIPs on Services due to its risk, so building more on top of it may be counterproductive , we event have an admission controller to deny them https://github.com/kubernetes/kubernetes/pull/97395

another thing we should take into consideration is the mutability of Services, so what happens if I do

ClusterIP + ExternalIP + ExternalTrafficPolicy:Local and I update the service to remove ExternalIP?

do I end with ClusterIP + ExternalTrafficPolicy:Local?

tnqn commented:
Just to confirm: should I close the PR?

thockin commented:
We either do this PR or we fix the comments that say policy applies to externalIPs.  The problem is that it DOES apply when type=LB and the rest of the time it works by dumb luck (we assume that if it is not Local it msut be Cluster, even though if we read it we would find "").

So I lean towards doing it.  What do you think @danwinship ?

thockin commented:
/approve

/hold I give the final LGTM or NAK to @danwinship (unless he wants to punt it back).

danwinship commented on pkg/api/v1/service/util.go:
```diff
@@ -70,7 +70,8 @@ func GetLoadBalancerSourceRanges(service *v1.Service) (utilnet.IPNetSet, error)
 // ExternalPolicyLocal checks if service has ETP = Local.
 func ExternalPolicyLocal(service *v1.Service) bool {
```
The doc comment is incorrect; the function checks both (a) if the service has ETP=Local, _and_ (b) if the service is _allowed_ to have ETP=local. This is kind of weird... (This is a pre-existing problem with the function.)

The registry/storage code should be validating the "allowed to have ETP=local" part itself anyway, so I think this function should _just_ check `service.Spec.ExternalTrafficPolicy` (though in that case it's not clear that we even need this function).

Alternatively/additionally, we could add an `ExternallyReachable` function that checks if it is NodePort or LB or has ExternalIPs, and then use that in a bunch of places.

danwinship commented on pkg/api/v1/service/util.go:
```diff
@@ -70,7 +70,8 @@ func GetLoadBalancerSourceRanges(service *v1.Service) (utilnet.IPNetSet, error)
 	if service.Spec.Type != v1.ServiceTypeLoadBalancer &&
-		service.Spec.Type != v1.ServiceTypeNodePort {
+		service.Spec.Type != v1.ServiceTypeNodePort &&
+		(service.Spec.Type != v1.ServiceTypeClusterIP || len(service.Spec.ExternalIPs) == 0) {
```
You don't need to check `ClusterIP` specifically; presumably we wany any type of service that is allowed to have `ExternalIPs` to be able to have `ExternalTrafficPolicy`.

danwinship commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -125,7 +125,8 @@ func SetDefaults_Service(obj *v1.Service) {
 	if (obj.Spec.Type == v1.ServiceTypeNodePort ||
-		obj.Spec.Type == v1.ServiceTypeLoadBalancer) &&
+		obj.Spec.Type == v1.ServiceTypeLoadBalancer ||
+		(obj.Spec.Type == v1.ServiceTypeClusterIP && len(obj.Spec.ExternalIPs) > 0)) &&
```
As above (and again in other places below), you only need to check "has ExternalIPs", not "is ClusterIP and has ExternalIPs".

The comment above is now slightly incorrect. I suggest "for externally-reachable services"

danwinship commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5219,7 +5219,7 @@ func validateServiceExternalTrafficPolicy(service *core.Service) field.ErrorList
 		if service.Spec.ExternalTrafficPolicy != "" {
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("externalTrafficPolicy"), service.Spec.ExternalTrafficPolicy,
-				"may only be set when `type` is 'NodePort' or 'LoadBalancer'"))
+				"may only be set when `type` is 'NodePort' or 'LoadBalancer', or when `type` is `ClusterIP` and `externalIPs` is set"))
```
"may only be set for externally-reachable services"?

tnqn replied:
updated

danwinship reviewed (commented):
I'm not sure I love it...

danwinship commented:
@thockin feel free to merge if you want to get it in for 1.28 freeze (and maybe have fixups after)

thockin commented:
I am AFK - it's your call.  Not urgent.

WRT testing clusterIP - I told them to do that because (sadly) we allow
ExternalIPs with ExternalName, but we don't DO anything with them.  I sent
a PR to you to add warnings.

On Tue, Jul 18, 2023, 9:37 AM Dan Winship ***@***.***> wrote:

> @thockin <https://github.com/thockin> feel free to merge if you want to
> get it in for 1.28 freeze (and maybe have fixups after)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/119150#issuecomment-1640564642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVC4RFK6BYOVJDWWQJLXQ23U5ANCNFSM6AAAAAA2BRV2PE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tnqn reviewed (commented):
@danwinship thanks for the comments. I have addressed all comments.

aojea commented:
> ClusterIP + ExternalIP + ExternalTrafficPolicy:Local and I update the service to remove ExternalIP?
> 
> do I end with ClusterIP + ExternalTrafficPolicy:Local?

can we please add an integration test that perform some of these mutations ?

you have some boilerplate here you can reuse https://github.com/kubernetes/kubernetes/blob/master/test/integration/service/service_test.go

tnqn commented:
> > ClusterIP + ExternalIP + ExternalTrafficPolicy:Local and I update the service to remove ExternalIP?
> > do I end with ClusterIP + ExternalTrafficPolicy:Local?
> 
> can we please add an integration test that perform some of these mutations ?
> 
> you have some boilerplate here you can reuse https://github.com/kubernetes/kubernetes/blob/master/test/integration/service/service_test.go

Thanks for the link. I have updated related func and added an integration test to validate ExternalTrafficPolicy will be wiped when it's no longer needed.

aojea commented on test/integration/service/service_test.go:
```diff
@@ -180,3 +180,64 @@ func Test_ConvertingToExternalNameServiceDropsInternalTrafficPolicy(t *testing.T
+
+	if service.Spec.ExternalTrafficPolicy != "" {
+		t.Errorf("service externalTrafficPolicy should be droppped but is set: %v", service.Spec.ExternalTrafficPolicy)
+	}
```
sorry for being picky, what if we add a new operation here to Update ExternalIPs to IPx and check the ExternalTrafficPolicy comes back? I think this will cover the whole cycle

tnqn commented:
/retest

tnqn commented:
/retest

danwinship commented on pkg/api/service/util_test.go:
```diff
@@ -129,6 +129,43 @@ func TestAllowAll(t *testing.T) {
+		if res != expect {
+			t.Errorf("Expected ExternallyAccessible = %v, got %v", expect, res)
+		}
+	}
```
call `t.Helper()` as the first line of `checkExternallyAccessible`, so that if `t.Errorf()` gets called, it will report that the failure occurred on the line number of the call to `checkExternallyAccessible()`, rather than always claiming the error occurred on line 136.

EDIT: oh, I see, you copied this pattern from other tests... maybe consider a follow-up PR to fix all of them...

danwinship commented on pkg/api/service/util_test.go:
```diff
@@ -129,6 +129,43 @@ func TestAllowAll(t *testing.T) {
+	checkExternallyAccessible(false, &api.Service{
+		Spec: api.ServiceSpec{
+			Type: api.ServiceTypeExternalName,
+		},
```
you should add a test for ExternalName + non-empty ExternalIPs too, since that's the really non-obvious one

tnqn replied:
added

danwinship commented on pkg/api/v1/service/util_test.go:
```diff
@@ -129,6 +129,43 @@ func TestAllowAll(t *testing.T) {
+	checkExternallyAccessible(false, &v1.Service{
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeExternalName,
+		},
```
add the ExternalName+ExternalIPs test here too

tnqn replied:
added

danwinship commented on test/integration/service/service_test.go:
```diff
@@ -120,7 +120,7 @@ func Test_ExternalNameServiceDropsInternalTrafficPolicy(t *testing.T) {
 // Test_ConvertingToExternalNameServiceDropsInternalTrafficPolicy tests that converting a Service to Type=ExternalName
 // results in the internalTrafficPolicy field being dropped.This test exists due to historic reasons where the internalTrafficPolicy
-// field was being defaulted in older versions. New versions stop defauting the field and drop on read, but for compatibility reasons
+// field was being defaulted in older versions. New versions stop defaulting the field and drop on read, but for compatibility reasons
```
(I would put these typo/error message text fixes into a separate commit, so that if it turns out that this PR breaks things and we have to revert it, we can revert the "Allow specifying ExternalTrafficPolicy for ClusterIP Services with ExternalIPs" commit, but still keep the "fix typos in comments" commit.)

tnqn replied:
makes sense, moved typo fixes to a separate commit.

danwinship reviewed (approved):
Just test nitpicks. I can't fully /approve this PR anyway because of the `pkg/api` stuff

danwinship commented:
oh, Tim already approved it. So yeah, just address those comments and I'll lgtm it. (Feel free to poke me on Slack to re-review.)

tnqn committed:
Fix typos in comments

Signed-off-by: Quan Tian <qtian@vmware.com>

tnqn reviewed (commented):
Thanks @danwinship, I have addressed all comments.

danwinship commented:
/lgtm
thanks!

tnqn commented:
Thanks @danwinship. There is a hold label added in https://github.com/kubernetes/kubernetes/pull/119150#issuecomment-1635158797, could you please unhold it?

danwinship commented:
/hold cancel
